### ISSUE TITLE
How do I obtain the installation INSTALLDIR in a custom action on uninstall?

### ISSUE THREAD

**Issue by NeloBorcram** [oleg-shilo/wixsharp#486]
I would need to execute a file on uninstall before it's removed by the installer.
The problem I am facing is that it works just fine as long as I am not changing the default path but using the custom INSTALLDIR path during setup but once I change it to a default path the whole uninstall still seems to work (like removing all files and stuff) but the custom action looks for the file on the default path.

```
var dir = new InstallDir(@"%ProgramFiles%\MyCompany\MyProduct",
                             new Files(@"..\..\..\AllMyFiles\*.*"));

var project = new Project("MyProduct", dir) {
     GUID = new Guid("ecfa8a2d-4887-40f0-8806-c437ac71dce8"),
     Version = new Version(1, 0, 0, 0),
     UI = WUI.WixUI_InstallDir,
     Platform = Platform.x64
};

project.Actions = new WixSharp.Action[] {
     new ElevatedManagedAction(CustomActions.InstallService,Return.check, When.Before, Step.InstallFinalize, Condition.NOT_Installed),
     new ElevatedManagedAction(CustomActions.UninstallService,Return.check, When.After, Step.PreviousAction, Condition.Installed)
};

```

```
[CustomAction]
public static ActionResult StartService(Session session) {

     string installDir = session.Property("INSTALLDIR"); //<--this works on install even when using a custom path
     string workingDir = Path.Combine(installDir, @"\SomePathToTheBatchFile");
     RunCmdMethode(workingDir, "something.bat -some arguments");

     return ActionResult.Success;
}
[CustomAction]
public static ActionResult UninstallService(Session session) {

     string installDir = session.Property("INSTALLDIR"); //<--this does not give back the right path on uninstall in case the default path was changed during installation
     string workingDir = Path.Combine(installDir, @"\SomePathToTheBatchFile");
     RunCmdMethode(workingDir, "something.bat -some arguments");

     return ActionResult.Success;
}
```

Any hints would be appreciated. Thank you!

**Comment by oleg-shilo** [comment]
Scheduling custom actions is a convoluted process that can easily be done wrong. Most likely if you replace `Step.PreviousAction` with something more appropriate the problem will go away.

Though, if you ask me, nothing beats simplicity and intuitive nature of C# events:

```C#
var project =
        new ManagedProject("My Product",
            new Dir(@"%ProgramFiles%\My Product",
                new File("myapp.exe",
                . . .
        
project.UI = WUI.WixUI_InstallDir;

project.BeforeInstall += args =>
{
    if (args.IsUninstalling)
        RunCmdMethode(args.InstallDir.PathCombine(SomePathToTheBatchFile), 
                     "something.bat -uninstall");
    else
        RunCmdMethode(args.InstallDir.PathCombine(SomePathToTheBatchFile), 
                     "something.bat -install");
};

project.BuildMsi();
```

**Comment by NeloBorcram** [comment]
Thank you very much for your fast reply.

Regarding `Step.PreviousAction` I actually made a mistake when posting the code here since I wanted to show an excerpt of the whole code to not fill the page unnecessarily in the full code I have another uninstall action before. That way it actually makes sense:
```
new ElevatedManagedAction(CustomActions.InstallService, Return.check, When.After, Step.PreviousAction, Condition.NOT_Installed),
new ElevatedManagedAction(CustomActions.StartService, Return.check, When.After, Step.PreviousAction, Condition.NOT_Installed),
new ElevatedManagedAction(CustomActions.StopService, Return.check, When.Before, Step.RemoveFiles, Condition.Installed),
new ElevatedManagedAction(CustomActions.UninstallService, Return.check, When.After, Step.PreviousAction, Condition.Installed)
```
I am sorry for not being more careful with the pasted code.
Sadly the custom actions require elevated permissions so I guess I have to stick to the `ElevatedManagedAction` approach although I would be happy to use the Event approach. Thank you for the tip.

I finally could solve the problem. Since `session.Property("INSTALLDIR")` should actually work and so I did not make a mistake at that point, I could figure out the root cause, which is setting the `IsInstallDir` property to true by using the `InstallDir` class instead of the `Dir` class. It overwrote the `INSTALLDIR` property when uninstalling back to the hard coded default path.
This explains why the setup worked fine as long as using the default path and also why it worked for all the install custom steps even when using a custom path but not for uninstalling anymore. The reason tho, why I set the `IsInstallDir` property to true in the first place is because of some weird behavior when adding all the files to the setup using wildcards. As long as there are multiple files and folders in the source directory it would work just as expected, getting all paths right and so on. But once the source folder only contains a single folder inside which then contains the rest of the setup files within, it sets the inner folder to be the new root folder (is there a reason for that?) and so screws up many necessary paths. Using `InstallDir` instead of `Dir` fixed that.
I might put some work into restructuring the whole thing (if this is even possible in my use case), but for now simply adding a readme file on the same level as the single inner folder solves it.

Thank you very much for your time and effort.
